module/apmpgx/tracer.go

```diff
@@ -0,0 +1,162 @@
+
+	//querySpanType is setting action for batch expression trace in APM server.
+	batchSpanType = "db.postgresql.batch"
+
+	//postgresql is subtype which indicates database type in trace.
+	postgresql = "postgresql"
+)
+
+// ErrUnsupportedPgxVersion is indicating that data doesn't contain value for "time" key.
+// This fields appeared in pgx v4.17
+var ErrUnsupportedPgxVersion = errors.New("this version of pgx is unsupported, please upgrade to v4.17")
+
+// tracer is an implementation of pgx.Logger.
+type tracer struct {
+	// logger is the pgx.Logger to use for writing data to log.
+	// If logger is nil, then data won't be written to log, and only spans will be created.
+	logger pgx.Logger
+}
+
+// NewTracer returns a new tracer which creates spans for pgx queries.
+// It is safe to pass nil logger to constructor.
+func NewTracer(logger pgx.Logger) *tracer {
+	return &tracer{logger: logger}
+}
+
+// Log is getting type of SQL expression from msg and run suitable trace.
+// If logger was provided in NewTracer constructor, than expression will be
+// written to your logger that implements pgx.Logger interface.
+func (t *tracer) Log(ctx context.Context, level pgx.LogLevel, msg string, data map[string]interface{}) {
+	if t.logger != nil {
+		t.logger.Log(ctx, level, msg, data)
+	}
+
+	switch msg {
+	case "Query", "Exec":
+		t.QueryTrace(ctx, data)
+	case "CopyFrom":
+		t.CopyTrace(ctx, data)
+	case "SendBatch":
+		t.BatchTrace(ctx, data)
+	}
+}
+
+// QueryTrace traces query and creates spans for them.
+func (t *tracer) QueryTrace(ctx context.Context, data map[string]interface{}) {
+	stop := time.Now()
+
+	if _, ok := data["time"]; !ok {
+		apm.CaptureError(ctx, ErrUnsupportedPgxVersion).Send()
+		return
+	}
+
+	span, _ := apm.StartSpanOptions(ctx, apmsql.QuerySignature(data["sql"].(string)), querySpanType, apm.SpanOptions{
+		Start: stop.Add(-data["time"].(time.Duration)),
+	})
+
+	span.Duration = data["time"].(time.Duration)
+	span.Context.SetDatabase(apm.DatabaseSpanContext{
+		Type:      postgresql,
+		Statement: data["sql"].(string),
+	})
+
+	if _, ok := data["err"]; ok {
+		e := apm.CaptureError(ctx, data["err"].(error))
+		e.Timestamp = stop
+		e.Send()
+	}
+
+	span.End()
+}
+
+// CopyTrace traces copy queries and creates spans for them.
+func (t *tracer) CopyTrace(ctx context.Context, data map[string]interface{}) {
+	stop := time.Now()
+
+	if _, ok := data["time"]; !ok {
+		apm.CaptureError(ctx, ErrUnsupportedPgxVersion).Send()
+		return
+	}
+
+	span, _ := apm.StartSpanOptions(ctx, fmt.Sprintf("COPY TO %s", strings.Join(data["tableName"].(pgx.Identifier), ", ")),
+		copySpanType, apm.SpanOptions{
+			Start: stop.Add(-data["time"].(time.Duration)),
+		})
+
+	span.Duration = data["time"].(time.Duration)
+	span.Context.SetDatabase(apm.DatabaseSpanContext{
+		Type: postgresql,
+	})
+
+	if _, ok := data["err"]; ok {
+		e := apm.CaptureError(ctx, data["err"].(error))
+		e.Timestamp = stop
+		e.Send()
+	}
+
+	span.End()
```

Comment:
Most of the code is duplicated for each of the three methods - the main that differs is the span name and type.

Could you please take a look at extracting the common code into a separate method? e.g.

```go
// BatchTrace traces batch execution and creates spans for the whole batch.
func (t *tracer) BatchTrace(ctx context.Context, data map[string]interface{}) {
	span, ok := t.newSpan(ctx, "BATCH", batchSpanType)
	if !ok {
		return
	}
	defer span.End()

	if batchLen, ok := data["batchLen"].(int); ok {
		span.Context.SetLabel("batch.length", batchLen)
	}
}

func (t *tracer) newSpan(ctx context.Context, spanName, spanType string) (*apm.Span, bool) {
	stop := time.Now()

	duration, ok := data["time"].(time.Duration)
	if !ok {
		apm.CaptureError(ctx, ErrUnsupportedPgxVersion).Send()
		return nil, false
	}

	span, _ := apm.StartSpanOptions(ctx, spanName, spanType, apm.SpanOptions{
		Start: stop.Add(-duration),
		ExitSpan: true,
	})
	if span.Dropped() {
		span.End()
		return nil, false
	}

	span.Duration = duration
	span.Context.SetDatabase(apm.DatabaseSpanContext{
		Type: postgresql,
	})

	if err, ok := data["err"].(error); ok {
		e := apm.CaptureError(ctx, err)
		e.Timestamp = stop
		e.SetSpan(span)
		e.Send()
	}

	return span, true
}
```